src/catalog.jl

```diff
@@ -118,14 +130,136 @@ catalog = catalog_create_rest("http://polaris:8181")
 ```
 """
 function catalog_create_rest(uri::String; properties::Dict{String,String}=Dict{String,String}())
+    # Create an empty catalog (no authenticator)
+    catalog_ptr = @ccall rust_lib.iceberg_catalog_init()::Ptr{Cvoid}
+    if catalog_ptr == C_NULL
+        throw(IcebergException("Failed to create empty catalog"))
+    end
+
+    # Initialize the catalog with REST connection
+    # Convert properties dict to array of PropertyEntry structs
+    property_entries = [PropertyEntry(pointer(k), pointer(v)) for (k, v) in properties]
+    properties_len = length(property_entries)
+
     response = CatalogResponse()
 
+    async_ccall(response, property_entries, properties) do handle
+        @ccall rust_lib.iceberg_rest_catalog_create(
+            catalog_ptr::Ptr{Cvoid},
+            uri::Cstring,
+            (properties_len > 0 ? pointer(property_entries) : C_NULL)::Ptr{PropertyEntry},
+            properties_len::Csize_t,
+            response::Ref{CatalogResponse},
+            handle::Ptr{Cvoid}
+        )::Cint
+    end
+
+    @throw_on_error(response, "catalog_create_rest", IcebergException)
+
+    return Catalog(response.catalog, nothing)
+end
+
+"""
+    catalog_create_rest(authenticator::Function, uri::String; properties::Dict{String,String}=Dict{String,String}())::Catalog
+
+Create a REST catalog connection with custom token authentication.
+
+# Arguments
+- `authenticator::Function`: A callable that takes no arguments and returns a token string.
+  The function will be called whenever a new token is needed for authentication.
+- `uri::String`: URI of the Iceberg REST catalog server (e.g., "http://localhost:8181")
+- `properties::Dict{String,String}`: Optional key-value properties for catalog configuration.
+  By default (empty dict), no additional properties are passed.
+
+# Returns
+- A `Catalog` handle for use in other catalog operations
+
+# Example
+```julia
+function get_token()
+    return ENV["ICEBERG_TOKEN"]
+end
+
+catalog = catalog_create_rest(get_token, "http://polaris:8181")
+```
+"""
+function catalog_create_rest(authenticator::Function, uri::String; properties::Dict{String,String}=Dict{String,String}())
+    # Step 1: Create an empty catalog
+    catalog_ptr = @ccall rust_lib.iceberg_catalog_init()::Ptr{Cvoid}
+    if catalog_ptr == C_NULL
+        throw(IcebergException("Failed to create empty catalog"))
+    end
+
+    # Step 2: Wrap the authenticator in a mutable container so we can get a stable pointer
+    authenticator_ref = Ref(authenticator)
+
+    # Step 3: Create the C callback function that wraps the Julia authenticator
+    c_callback = @cfunction(
+        $(
+            function token_callback(user_data::Ptr{Cvoid}, token_ptr::Ptr{Ptr{Cchar}})::Cint
```

Comment:
Doesn't this have to be marked with Base.@ccallable?

Reply:
Shouldn't we still mark this as ccallable? I'm not sure I understand Julia here, but in other places we do mark callbacks like that.